aexpect/remote_door.py

```diff
@@ -885,3 +943,52 @@ def recreate_exception(class_name, class_dict):
     for exception in exceptions:
         # noinspection PyUnresolvedReferences
         Pyro4.util.SerializerBase.register_dict_to_class(exception, recreate_exception)
+
+
+def expose_remote_classes(classes=None, modules=None):
+    """
+    Make accessible all remote custom classes.
+
+    :param classes: full module path class names (optional)
+    :type classes: [str] or None
+    :param modules: full module paths whose custom classes will first be
+                    detected and then automatically exposed inclusive of
+                    parent classes and inheritance (optional)
+    :type modules: [str] or None
+    """
+    def list_module_classes(modstr):
+        imported_module = importlib.import_module(modstr)
+        module_classes = []
+        for name in imported_module.__dict__:
+            if not inspect.isclass(imported_module.__dict__[name]):
+                continue
+            module_classes.append(modstr + "." + name)
+        return module_classes
+
+    classes = [] if not classes else classes
+    modules = [] if not modules else modules
+    for module in modules:
+        classes += list_module_classes(module)
+    LOG.debug("Exposing the following classes (with proper inheritance): %s",
+              ", ".join(classes))
+
+    def get_class_from_name(clsstr):
+        module_name, class_name = clsstr.rsplit('.', 1)
+        module = importlib.import_module(module_name)
+        cls = getattr(module, class_name)
+        return cls
+
+    for cls in map(get_class_from_name, classes):
+        # the inheritance is inclusive of the class itself
+        for base_cls in cls.__mro__:
+            if base_cls in (object, int, float, bool, str, tuple, frozenset):
+                # known immutable classes should be skipped
+                continue
+            if inspect.ismethoddescriptor(base_cls):
+                # method descriptors should be skipped
+                continue
+            try:
+                server.expose(base_cls)
+            except (TypeError, AttributeError) as error:
+                logging.warning("Additional class exposing error: %s", error)
```

Comment:
IIUC it's not `additional` but `parent` class, right?

Reply:
Not quite, the mro will also include the starting class, here the `cls` itself. So what we will add is the extra class *and* all its base methods to provide a proper "network inheritance".

Reply:
Sure, I meant the same but wasn't explicit. My point is that `additional` seems like some extra added classes while here (IIUC) it's more about all classes that this class uses (so recursive parents). If so I think the message should be `Parent(s) class exponsing error` (or `Upper`?) rather than `Additional`. Again, take this just for consideration, no need to change it if you think `Additional` is more suitable. I'll definitely trust your opinion here since you wrote all of it.

Reply:
The `Parent` class exposing would be incorrect since we are exposing the initially given classes together with their parents. "Additional" here is in the sense that we don't just expose the class (or emulated module class) of the initial object but additional classes that it makes use of for which we will have to be explicit now. Thankfully these will be fewer cases with type hint annotated code overtaking more software projects in the future. The reason for this is that the above extra classes are also needed in addition to any classes that are either *defined or imported* in the original namespace, all of which will already be handled without this extra argument. Would you like me to document any of this further in the docstrings?

Reply:
Nop, I was just curious about the error message but it seems you had a good reason for this name :-)